File: sdks/csharp/src/Nanoservice.Core/Validation/SchemaValidator.cs
using System.Text.Json;

namespace Nanoservice.Core.Validation;

/// <summary>
/// SchemaValidator validates a JsonElement against a JSON Schema (Draft 7 subset).
/// Supports: type, required, properties, enum, minLength, maxLength, minimum, maximum.
/// </summary>
public class SchemaValidator
{
    /// <summary>
    /// Validate data against a JSON Schema. Returns a list of error messages.
    /// An empty list means validation passed.
    /// </summary>
    public List<string> Validate(JsonElement data, JsonElement schema)
    {
        var errors = new List<string>();
        ValidateValue(data, schema, "$", errors);
        return errors;
    }

    private void ValidateValue(JsonElement data, JsonElement schema, string path, List<string> errors)
    {
        if (schema.ValueKind != JsonValueKind.Object)
            return;

        // Type check
        if (schema.TryGetProperty("type", out var typeVal) && typeVal.ValueKind == JsonValueKind.String)
        {
            var expectedType = typeVal.GetString()!;
            if (!CheckType(data, expectedType))
            {
                errors.Add($"{path}: expected type \"{expectedType}\", got {TypeName(data)}");
                return;
            }
        }

        // Enum check
        if (schema.TryGetProperty("enum", out var enumVal) && enumVal.ValueKind == JsonValueKind.Array)
        {
            bool found = false;
            foreach (var item in enumVal.EnumerateArray())
            {
                if (JsonElementEquals(data, item))
                {
                    found = true;
                    break;
                }
            }
            if (!found)
            {
                errors.Add($"{path}: value not in allowed enum values");
            }
        }

        // Object: required fields
        if (data.ValueKind == JsonValueKind.Object &&
            schema.TryGetProperty("required", out var requiredVal) &&
            requiredVal.ValueKind == JsonValueKind.Array)
        {
            foreach (var field in requiredVal.EnumerateArray())
            {
                var fieldName = field.GetString();
                if (fieldName is not null && !data.TryGetProperty(fieldName, out _))
                {
                    errors.Add($"{path}: missing required field \"{fieldName}\"");
                }
            }
        }

        // Object: properties
        if (data.ValueKind == JsonValueKind.Object &&
            schema.TryGetProperty("properties", out var propertiesVal) &&
            propertiesVal.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in propertiesVal.EnumerateObject())
            {
                if (data.TryGetProperty(prop.Name, out var propData))
                {
                    var propPath = $"{path}.{prop.Name}";
                    ValidateValue(propData, prop.Value, propPath, errors);
                }
            }
        }

        // String constraints
        if (data.ValueKind == JsonValueKind.String)
        {
            var str = data.GetString() ?? string.Empty;

            if (schema.TryGetProperty("minLength", out var minLen) && minLen.TryGetInt32(out var min))
            {
                if (str.Length < min)
                {
                    errors.Add($"{path}: string length {str.Length} is less than minimum {min}");
                }
            }

            if (schema.TryGetProperty("maxLength", out var maxLen) && maxLen.TryGetInt32(out var max))
            {
                if (str.Length > max)
                {
                    errors.Add($"{path}: string length {str.Length} exceeds maximum {max}");
                }
            }
        }

        // Numeric constraints
        if (data.ValueKind == JsonValueKind.Number)
        {
            var num = data.GetDouble();

            if (schema.TryGetProperty("minimum", out var minVal) && minVal.ValueKind == JsonValueKind.Number)
            {
                var minimum = minVal.GetDouble();
                if (num < minimum)
                {
                    errors.Add($"{path}: value {num} is less than minimum {minimum}");
                }
            }

            if (schema.TryGetProperty("maximum", out var maxVal) && maxVal.ValueKind == JsonValueKind.Number)
            {
                var maximum = maxVal.GetDouble();
                if (num > maximum)
                {
                    errors.Add($"{path}: value {num} exceeds maximum {maximum}");
                }
            }
        }

        // Array items
        if (data.ValueKind == JsonValueKind.Array &&
            schema.TryGetProperty("items", out var itemsSchema))
        {
            int i = 0;
            foreach (var item in data.EnumerateArray())
            {
                var itemPath = $"{path}[{i}]";
                ValidateValue(item, itemsSchema, itemPath, errors);
                i++;
            }
        }
    }

    private static bool CheckType(JsonElement data, string expectedType)
    {
        return expectedType switch
        {
            "string" => data.ValueKind == JsonValueKind.String,
            "number" => data.ValueKind == JsonValueKind.Number,
            "integer" => data.ValueKind == JsonValueKind.Number && IsInteger(data),
            "boolean" => data.ValueKind == JsonValueKind.True || data.ValueKind == JsonValueKind.False,
            "object" => data.ValueKind == JsonValueKind.Object,
            "array" => data.ValueKind == JsonValueKind.Array,
            "null" => data.ValueKind == JsonValueKind.Null,
            _ => true
        };
    }

    private static bool IsInteger(JsonElement data)
    {
        if (data.TryGetInt64(out _)) return true;
        var d = data.GetDouble();
        return Math.Abs(d % 1) < double.Epsilon;
    }

    private static string TypeName(JsonElement data)
    {
        return data.ValueKind switch
        {
            JsonValueKind.String => "string",
            JsonValueKind.Number => "number",
            JsonValueKind.True => "boolean",
            JsonValueKind.False => "boolean",
            JsonValueKind.Object => "object",
            JsonValueKind.Array => "array",
            JsonValueKind.Null => "null",
            JsonValueKind.Undefined => "undefined",
            _ => "unknown"
        };
    }

    private static bool JsonElementEquals(JsonElement a, JsonElement b)
    {
        if (a.ValueKind != b.ValueKind) return false;

        return a.ValueKind switch
        {
            JsonValueKind.String => a.GetString() == b.GetString(),
            JsonValueKind.Number => a.GetDouble() == b.GetDouble(),
            JsonValueKind.True => true,
            JsonValueKind.False => true,
            JsonValueKind.Null => true,
            _ => a.GetRawText() == b.GetRawText()
        };
    }
}

File: sdks/java/src/main/java/com/blok/nanoservice/validation/SchemaValidator.java
package com.blok.nanoservice.validation;

import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Lightweight JSON Schema validator supporting a subset of JSON Schema Draft 7.
 * <p>
 * Supported keywords: type, required, properties, enum, minLength, maxLength,
 * minimum, maximum.
 */
public class SchemaValidator {

    private static final Gson GSON = new Gson();

    /**
     * Validates data against a JSON Schema.
     *
     * @param data   the data to validate
     * @param schema the JSON Schema as a map
     * @return a list of validation errors (empty if valid)
     */
    public List<String> validate(Object data, Map<String, Object> schema) {
        List<String> errors = new ArrayList<>();
        if (schema == null) {
            return errors;
        }
        validateValue(data, schema, "", errors);
        return errors;
    }

    @SuppressWarnings("unchecked")
    private void validateValue(Object data, Map<String, Object> schema, String path, List<String> errors) {
        if (schema == null) {
            return;
        }

        // Type check
        Object schemaType = schema.get("type");
        if (schemaType instanceof String typeStr) {
            if (!checkType(data, typeStr)) {
                errors.add(pathStr(path) + ": expected type \"" + typeStr + "\", got " + describeType(data));
                return;
            }
        }

        // Enum check
        Object enumVal = schema.get("enum");
        if (enumVal instanceof List<?> enumList) {
            if (!checkEnum(data, enumList)) {
                errors.add(pathStr(path) + ": value not in allowed enum values");
            }
        }

        // Object properties
        Object propsVal = schema.get("properties");
        if (propsVal instanceof Map<?, ?> props && data instanceof Map<?, ?> dataMap) {
            for (Map.Entry<?, ?> entry : props.entrySet()) {
                String propName = String.valueOf(entry.getKey());
                if (entry.getValue() instanceof Map<?, ?> propSchema) {
                    String propPath = path.isEmpty() ? propName : path + "." + propName;
                    Object propData = dataMap.get(propName);
                    if (propData != null) {
                        validateValue(propData, (Map<String, Object>) propSchema, propPath, errors);
                    }
                }
            }
        }

        // Required fields
        Object requiredVal = schema.get("required");
        if (requiredVal instanceof List<?> requiredList && data instanceof Map<?, ?> dataMap) {
            for (Object req : requiredList) {
                String fieldName = String.valueOf(req);
                if (!dataMap.containsKey(fieldName)) {
                    errors.add(pathStr(path) + ": missing required field \"" + fieldName + "\"");
                }
            }
        }

        // String constraints
        if (data instanceof String str) {
            Object minLenVal = schema.get("minLength");
            if (minLenVal instanceof Number minLen) {
                if (str.length() < minLen.intValue()) {
                    errors.add(pathStr(path) + ": string length " + str.length()
                            + " is less than minimum " + minLen.intValue());
                }
            }
            Object maxLenVal = schema.get("maxLength");
            if (maxLenVal instanceof Number maxLen) {
                if (str.length() > maxLen.intValue()) {
                    errors.add(pathStr(path) + ": string length " + str.length()
                            + " exceeds maximum " + maxLen.intValue());
                }
            }
        }

        // Numeric constraints
        Double num = toDouble(data);
        if (num != null) {
            Object minVal = schema.get("minimum");
            if (minVal instanceof Number minimum) {
                if (num < minimum.doubleValue()) {
                    errors.add(pathStr(path) + ": value " + num
                            + " is less than minimum " + minimum.doubleValue());
                }
            }
            Object maxVal = schema.get("maximum");
            if (maxVal instanceof Number maximum) {
                if (num > maximum.doubleValue()) {
                    errors.add(pathStr(path) + ": value " + num
                            + " exceeds maximum " + maximum.doubleValue());
                }
            }
        }
    }

    private boolean checkType(Object data, String expectedType) {
        if (data == null) {
            return "null".equals(expectedType);
        }
        return switch (expectedType) {
            case "string" -> data instanceof String;
            case "number" -> data instanceof Number;
            case "integer" -> {
                if (data instanceof Number num) {
                    double d = num.doubleValue();
                    yield d == Math.floor(d) && !Double.isInfinite(d);
                }
                yield false;
            }
            case "boolean" -> data instanceof Boolean;
            case "object" -> data instanceof Map;
            case "array" -> data instanceof List;
            case "null" -> false; // data is not null at this point
            default -> true;
        };
    }

    private boolean checkEnum(Object data, List<?> enumValues) {
        String dataJson = GSON.toJson(data);
        for (Object allowed : enumValues) {
            String allowedJson = GSON.toJson(allowed);
            if (dataJson.equals(allowedJson)) {
                return true;
            }
        }
        return false;
    }

    private Double toDouble(Object data) {
        if (data instanceof Number num) {
            return num.doubleValue();
        }
        return null;
    }

    private String describeType(Object data) {
        if (data == null) return "null";
        if (data instanceof String) return "string";
        if (data instanceof Boolean) return "boolean";
        if (data instanceof Number) return "number";
        if (data instanceof Map) return "object";
        if (data instanceof List) return "array";
        return data.getClass().getSimpleName();
    }

    private String pathStr(String path) {
        if (path == null || path.isEmpty()) {
            return "$";
        }
        return "$." + path;
    }
}
